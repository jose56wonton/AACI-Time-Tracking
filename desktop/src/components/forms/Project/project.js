import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Grid, Typography, Button, IconButton } from '@material-ui/core';
import cx from 'classnames';
import { Field, Form, FieldArray } from 'formik';
import { withStyles } from '@material-ui/core/styles';
import { Close } from '@material-ui/icons';

import TextField from 'components/inputs/TextField';
import Switch from 'components/inputs/Switch';
import Select from 'components/inputs/Select';
import styles from './styles';


export class Project extends Component {
  render() {

    const { classes, label, categories, subcategories, tasks, 
      isSubmitting, resetForm, initialValues, errors, values, type } = this.props;
 
    return (
      <Form>
        <Grid container spacing={24} className={classes.gridContainer}>
          <Grid item xs={12} className={cx(classes.headerRow, classes.row)}>
            <Typography variant="h6">{label}</Typography>          
          </Grid>
          <Grid item xs={12} className={classes.row}>
            <Field
              name="name"
              component={TextField}
              margin="none"
              label="Project Name"
              type="search"
              className={classes.field}
              helper="normal"
            />
            {[`add`,`edit`].includes(type) && (
              <Field
                name="date"
                component={TextField}
                margin="none"
                label="Start Date"
                type="date"
                className={classes.field}
                helper="normal"
              />)}
            <Field
              name="isActive"
              component={Switch}
              label="Active"
              className={classes.field}
            />
          </Grid>
          { [`filter`].includes(type) && (
            <Grid item xs={12} className={classes.row}>            
              <Field
                name="startTime"
                component={TextField}
                margin="none"
                label="Start Date"
                type="date"
                className={classes.field}
                helper="normal"
              />
              <Field
                name="endTime"
                component={TextField}
                margin="none"
                label="End Date"
                type="date"
                className={classes.field}
                helper="normal"
              />
            </Grid>)}
          { [`add`,`edit`].includes(type) && (
            <FieldArray
              name="projectTasks"
              render={arrayHelpers => {

                return (
                  <Grid item xs={12} container className={classes.body}>
                    {values.projectTasks &&
                    values.projectTasks.map((projectTasks, index) => {
                      return (
                        <Grid item xs={12}
                          key={index}
                          className={cx(
                            classes.card,
                            classes.verticalCenterBox,
                          )}
                        >
                          <div className={cx(classes.row, classes.bodyRow)}>
                            <Field
                              name={`projectTasks.${index}.categoryId`}
                              component={Select}
                              items={categories}
                              id={`${PROJECT_FORM_CATEGORY_FIELD_ID}_${index}`}
                              fullWidth
                              label="Category"
                              className={classes.field}
                              onChange={() => {
                                arrayHelpers.form.setFieldValue(
                                  `projectTasks.${index}.subcategoryId`,
                                  -1,
                                );
                                arrayHelpers.form.setFieldValue(
                                  `projectTasks.${index}.taskId`,
                                  -1,
                                );
                              }}
                            />
                            <Field
                              name={`projectTasks.${index}.subcategoryId`}
                              component={Select}
                              items={subcategories.filter(subcat => {
                                return subcat.categoryId === projectTasks.categoryId;
                              })}
                              id={`${PROJECT_FORM_SUBCATEGORY_FIELD_ID}_${index}`}
                              fullWidth
                              label="Subcategory"
                              className={classes.field}
                              onChange={() => {
                                arrayHelpers.form.setFieldValue(
                                  `projectTasks.${index}.taskId`,
                                  -1,
                                );
                              }}
                            />
                            <Field
                              name={`projectTasks.${index}.taskId`}
                              component={Select}
                              items={tasks.filter(globalTask => {
                                return globalTask.subcategory.id === projectTasks.subcategoryId && globalTask.category.id === projectTasks.categoryId;
                              })}
                              fullWidth
                              label="Task"
                              className={classes.field}
                            />
                            <Field                              
                              name={`projectTasks.${index}.estimateTime`}
                              component={TextField}
                              fullWidth
                              label="Estimated Time"
                              className={classes.field}
                            />
                            <Field                              
                              name={`projectTasks.${index}.quantity`}
                              component={TextField}
                              fullWidth
                              label="Quantity"
                              className={classes.field}
                            />
                            <div className={classes.verticalCenter}>
                              <IconButton
                                type="button"
                                color="secondary"
                                className={classes.iconButton}
                                id={`${PROJECT_FORM_REMOVE_PROJECT_TASK_BUTTON_ID}_${index}`}
                                onClick={() => arrayHelpers.remove(index)}
                              >
                                <Close />
                              </IconButton>
                            </div>
                          </div>
                        </Grid>
                      );
                    })
                    }
                    <Grid item xs={12} className={cx(classes.row, classes.footerRow)}>
                      <div className={classes.lunchBox}>

                      </div>
                      <Button
                        color="primary"
                        variant="contained"
                        id={PROJECT_FORM_ADD_PROJECT_TASK_BUTTON_ID}
                        onClick={() =>
                          arrayHelpers.push({
                            categoryId: -1,
                            subcategoryId: -1,
                            taskId: -1,
                            quantity: 1,
                            estimateTime: 1
                          })
                        }
                      >
                      Add Task
                      </Button>
                    </Grid>
                  </Grid>
                );
              }
              }
            />
          )}
          <Grid item xs={12} className={classes.row}>
            <Typography
              color="error"
              variant="button"
              className={classes.field}
            >
              {errors.submit}
            </Typography>
            <div>
              <Button
                type="submit"
                id={PROJECT_FORM_SUBMIT_BUTTON_ID}
                color="primary"
                disabled={isSubmitting || Object.keys(errors).length !== 0 }
                variant="contained"
                className={classes.button}
              >
                {[`add`,`edit`].includes(type) ? `Save`: `Apply`}
              </Button>
              <Button
                onClick={() => {
                  resetForm(initialValues);
                }}
                id={PROJECT_FORM_RESET_BUTTON_ID}
                disabled={isSubmitting}
                color="secondary"
                variant="text"
                className={classes.button}
              >
                Reset
              </Button>
              {[`filter`].includes(type) && 
              <Button
                onClick={() => {
                  resetForm(initialValues);                  
                  this.props.clearFilter();
                }}
                id={PROJECT_FORM_CLEAR_BUTTON_ID}
                disabled={isSubmitting}
                color="secondary"
                variant="text"
                className={classes.button}
              >
                Clear
              </Button>}
              
            </div>
          </Grid>
        </Grid>
      </Form>
    );
  }
}

export const PROJECT_FORM_RESET_BUTTON_ID = `project_form_reset_button`;
export const PROJECT_FORM_CLEAR_BUTTON_ID = `project_form_clear_button`;
export const PROJECT_FORM_SUBMIT_BUTTON_ID = `project_form_submit_button`;
export const PROJECT_FORM_ADD_PROJECT_TASK_BUTTON_ID = `project_form_add_project_task_button`;
export const PROJECT_FORM_REMOVE_PROJECT_TASK_BUTTON_ID = `project_form_remove_project_task_button`;
export const PROJECT_FORM_CATEGORY_FIELD_ID = `project_form_category_field`;
export const PROJECT_FORM_SUBCATEGORY_FIELD_ID = `project_form_subcategory_field`;

Project.propTypes = {
  classes: PropTypes.object.isRequired,
  type: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  removeProject: PropTypes.func,
  resetForm: PropTypes.func.isRequired,
  initialValues: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired,
  categories: PropTypes.array.isRequired,
  subcategories: PropTypes.array.isRequired,
  values: PropTypes.object.isRequired,
  tasks: PropTypes.array.isRequired,
  clearFilter: PropTypes.func
};

export default withStyles(styles)(Project);