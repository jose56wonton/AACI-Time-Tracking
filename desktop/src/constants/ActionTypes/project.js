export const GET_PROJECTS_REQUEST = 'get_projects_request';
export const GET_PROJECTS_SUCCESS = 'get_projects_success';
export const GET_PROJECTS_FAILURE = 'get_projects_failure';

export const DELETE_PROJECT_REQUEST = 'delete_project_request';
export const DELETE_PROJECT_SUCCESS = 'delete_project_success';
export const DELETE_PROJECT_FAILURE = 'delete_project_failure';

export const CREATE_PROJECT_REQUEST = 'create_project_request';
export const CREATE_PROJECT_SUCCESS = 'create_project_success';
export const CREATE_PROJECT_FAILURE = 'create_project_failure';

export const UPDATE_PROJECT_REQUEST = 'update_project_request';
export const UPDATE_PROJECT_SUCCESS = 'update_project_success';
export const UPDATE_PROJECT_FAILURE = 'update_project_failure';