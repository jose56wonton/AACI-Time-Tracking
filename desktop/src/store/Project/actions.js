import { projectActionTypes } from 'constants/actionTypeConstants';

import endpoints from './endpoints';
import { snackActions, projectTaskActions,analyzeActions,genericActions } from 'store/actions';
import * as status from 'constants/status';
import domains from 'constants/domains';

export const getAllProjects = () => {
  return async dispatch => {
    dispatch(genericActions.getAll(domains.PROJECT));    
  };
};

export const createProject = project => {
  return async dispatch => {
    dispatch({ type: projectActionTypes.CREATE_PROJECT_REQUEST });
    try {
      const response = await dispatch(genericActions.post(domains.PROJECT,project));
      for(const projectTask of project.projectTasks){
        projectTask.projectId = response.data.id;
        await dispatch(projectTaskActions.postProjectTask(projectTask));
      }
      await dispatch(analyzeActions.selectProject(response.data.id));
      await dispatch(snackActions.openSnack(status.SUCCESS, 'Project Created'));
      return dispatch({ type: projectActionTypes.CREATE_PROJECT_SUCCESS });      
    } catch (e) {
      console.log(e);
      await dispatch(snackActions.openSnack(status.SUCCESS, 'Project Creation Failed'));
      return dispatch({ type: projectActionTypes.CREATE_PROJECT_FAILURE });
    }
  };
};

export const updateProject = project => {
  return async dispatch => {
    dispatch({ type: projectActionTypes.UPDATE_PROJECT_REQUEST });
    try {
      const response = await dispatch(genericActions.put(domains.PROJECT,project));
      const { data } = await endpoints.getProjectTasksByProjectId(project);
      for(const projectTaskInDb of data){
        if(!project.projectTasks.find(ele => ele.id === projectTaskInDb.id)){
          await dispatch(projectTaskActions.deleteProjectTask(projectTaskInDb));
        }
      }
      for(const projectTask of project.projectTasks){
        if(projectTask.id){
          projectTask.projectId = response.data.id;          
          await dispatch(projectTaskActions.putProjectTask(projectTask));
        }else{
          projectTask.projectId = response.data.id;          
          await dispatch(projectTaskActions.postProjectTask(projectTask));
        }        
      }
      await dispatch(snackActions.openSnack(status.SUCCESS, 'Project Updated'));
      return dispatch({ type: projectActionTypes.UPDATE_PROJECT_SUCCESS });      
    } catch (e) {
      console.log(e);
      await dispatch(snackActions.openSnack(status.SUCCESS, 'Project Update Failed'));
      return dispatch({ type: projectActionTypes.UPDATE_PROJECT_FAILURE });
    }
  };
};

export const removeProject = id => {
  return async dispatch => {
    dispatch({ type: projectActionTypes.REMOVE_PROJECT_REQUEST });
    try {
      await dispatch(analyzeActions.deleteSelected(domains.PROJECT));
      await dispatch(genericActions.delet(domains.PROJECT,id));

      await dispatch(snackActions.openSnack(status.SUCCESS, 'Project Deleted'));
      return dispatch({ type: projectActionTypes.REMOVE_PROJECT_SUCCESS });      
    } catch (e) {
      console.log(e);
      await dispatch(snackActions.openSnack(status.SUCCESS, 'Project Deletion Failed'));
      return dispatch({ type: projectActionTypes.REMOVE_PROJECT_FAILURE });
    }
  };
};

