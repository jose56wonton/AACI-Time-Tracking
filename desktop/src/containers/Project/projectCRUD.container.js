import React, { Component } from 'react';
import { connect } from 'react-redux';

import { Formik } from 'formik';
import { Typography } from '@material-ui/core';
import moment from 'moment';

import Hero from 'components/layouts/Hero';
import Project from 'components/forms/Project';

import {
  subcategorySelectors,
  categorySelectors,
  taskSelectors,
  projectTaskSelectors
} from 'store/selectors';
import { projectActions } from 'store/actions';

import { analyzeStatus } from 'constants/analyze';
import { projectValidation } from 'constants/formValidation';

export class ProjectCRUD extends Component {
  removeProject = () => {
    const { selected, removeProject } = this.props;
    removeProject(selected.id);
  };
  render() {
    const { selected, status, categories, subcategories, tasks } = this.props;
    if (status === analyzeStatus.INIT) {
      return (
        <Hero fullWidth fullHeight>
          <Typography variant="h6">Select a Project.. </Typography>
        </Hero>
      );
    }

    if (status === analyzeStatus.ADDING) {
      return (
        <Formik
          enableReinitialize
          initialValues={{
            name: ``,
            isActive: true,
            date: moment()
              .startOf(`day`)
              .format(`YYYY-MM-DD`),
            projectTasks: []
          }}
          validationSchema={projectValidation}
          onSubmit={(values, formikFunctions) => {
            const { createProject } = this.props;
            return createProject({
              name: values.name,
              isActive: values.isActive ? 1 : 0,
              date: moment(values.date).format(`MM-DD-YY HH:mm:ss`),
              projectTasks: values.projectTasks
            }).then(
              () => {
                formikFunctions.resetForm();
                formikFunctions.setStatus({ success: true });
              },
              e => {
                formikFunctions.setStatus({ success: false });
                formikFunctions.setSubmitting(false);
                formikFunctions.setErrors({ submit: e });
              }
            );
          }}
          render={formikProps => {
            return (
              <Project
                label="Add"
                type="add"
                categories={categories}
                subcategories={subcategories}
                tasks={tasks}
                {...formikProps}
              />
            );
          }}
        />
      );
    }

    if (status === analyzeStatus.EDITING) {
      return (
        <Formik
          enableReinitialize
          initialValues={{
            ...selected,
            isActive: selected.isActive ? true : false,
            date: moment
              .utc(selected.date)
              .local()
              .startOf(`day`)
              .format(`YYYY-MM-DD`),
            projectTasks: selected.projectTasks.map(projectTask => {
              return {
                ...projectTask,
                subcategoryId: projectTask.task.subcategoryId,
                categoryId: projectTask.task.category.id
              };
            })
          }}
          validationSchema={projectValidation}
          onSubmit={(values, formikFunctions) => {
            const { updateProject } = this.props;

            return updateProject({
              id: values.id,
              name: values.name,
              isActive: values.isActive ? 1 : 0,
              date: moment(values.date).format(`MM-DD-YY HH:mm:ss`),
              projectTasks: values.projectTasks
            }).then(
              () => {
                formikFunctions.resetForm();
                formikFunctions.setStatus({ success: true });
              },
              e => {
                formikFunctions.setStatus({ success: false });
                formikFunctions.setSubmitting(false);
                formikFunctions.setErrors({ submit: e });
              }
            );
          }}
          render={formikProps => {
            return (
              <Project
                label="Edit"
                type="edit"
                categories={categories}
                subcategories={subcategories}
                tasks={tasks}
                removeProject={this.removeProject}
                {...formikProps}
              />
            );
          }}
        />
      );
    }
  }
}

/* istanbul ignore next */
const mapStateToProps = state => {
  return {
    selected: projectTaskSelectors.getSelectedProject(state),
    status: state.analyze.projectStatus,
    categories: categorySelectors.getAllCategories(state),
    subcategories: subcategorySelectors.getAllSubcategories(state),
    tasks: taskSelectors.getAllTasksWithContent(state)
  };
};

/* istanbul ignore next */
const mapDispatchToProps = dispatch => {
  return {
    createProject: project => {
      return dispatch(projectActions.createProject(project));
    },
    updateProject: project => {
      return dispatch(projectActions.updateProject(project));
    },
    removeProject: id => {
      return dispatch(projectActions.removeProject(id));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProjectCRUD);
