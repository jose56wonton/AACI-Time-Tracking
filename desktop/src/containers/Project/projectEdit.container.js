import React, { useEffect } from 'react';
import moment from 'moment';
import { projectValidation } from 'constants/formValidation';
import ProjectEdit from 'components/forms/Project/projectEdit';
import { Formik } from 'formik';
import { useDispatch } from 'react-redux';
import { projectActions } from 'store/actions';
import axios from 'helpers/axios';

const ProjectEditContainer = ({
  selected,
  categories,
  subcategories,
  tasks
}) => {
  const dispatch = useDispatch();

  useEffect(() => {
    axios
      .get(`/projects/${selected.id}/projectTasks?filter[include][activities]`)
      .then(({ data }) => {
        console.log(data);
      });
  }, [selected]);

  const removeProject = () =>
    dispatch(projectActions.removeProject(selected.id));

  const updateProject = project =>
    dispatch(projectActions.updateProject(project));

  return (
    <Formik
      enableReinitialize
      initialValues={{
        ...selected,
        isActive: selected.isActive ? true : false,
        date: moment
          .utc(selected.date)
          .local()
          .startOf(`day`)
          .format(`YYYY-MM-DD`),
        projectTasks: selected.projectTasks.map(projectTask => {
          return {
            ...projectTask,
            subcategoryId: projectTask.task.subcategoryId,
            categoryId: projectTask.task.category.id
          };
        })
      }}
      validationSchema={projectValidation}
      onSubmit={(values, formikFunctions) => {
        return updateProject({
          id: values.id,
          name: values.name,
          isActive: values.isActive ? 1 : 0,
          date: moment(values.date).format(`MM-DD-YY HH:mm:ss`),
          projectTasks: values.projectTasks
        }).then(
          () => {
            formikFunctions.resetForm();
            formikFunctions.setStatus({ success: true });
          },
          e => {
            formikFunctions.setStatus({ success: false });
            formikFunctions.setSubmitting(false);
            formikFunctions.setErrors({ submit: e });
          }
        );
      }}
      render={formikProps => {
        return (
          <ProjectEdit
            categories={categories}
            subcategories={subcategories}
            tasks={tasks}
            removeProject={removeProject}
            {...formikProps}
          />
        );
      }}
    />
  );
};

export default ProjectEditContainer;
