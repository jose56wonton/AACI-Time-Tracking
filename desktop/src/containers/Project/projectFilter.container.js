import React, { Component } from 'react';
import { connect } from 'react-redux';

import { Formik } from 'formik';
import moment from 'moment';
import { Card } from '@material-ui/core';

import { subcategorySelectors,categorySelectors,taskSelectors } from 'store/selectors';
import { analyzeActions } from 'store/actions';
import ProjectForm from 'components/forms/Project';

import domain from 'constants/domains';

export class ProjectFilter extends Component {

  removeProject = () => {
    const { selected,removeProject } = this.props;
    removeProject(selected.id);
  }
  render () {
    const { projectFilters, projectFilterVisible, clearFilter } = this.props;
    
    if(projectFilterVisible){
      return (
        <Formik
          enableReinitialize
          initialValues={{
            ...projectFilters,        
            startTime: moment(projectFilters.startTime,`MM-DD-YY HH:mm:ss`).format(`YYYY-MM-DDTHH:mm`),
            endTime: moment(projectFilters.endTime,`MM-DD-YY HH:mm:ss`).format(`YYYY-MM-DDTHH:mm`)
          }}
          onSubmit={(values,formikFunctions) => {
            this.props.updateFilter({ 
              ...values,
              startTime: moment(values.startTime,`YYYY-MM-DDTHH:mm`).format(`MM-DD-YY HH:mm:ss`),
              endTime: moment(values.endTime,`YYYY-MM-DDTHH:mm`).format(`MM-DD-YY HH:mm:ss`)
            });
            formikFunctions.resetForm();  
          }}
          render={formikProps => {
            return (
              <Card style={{ position: `absolute`, top: `70px`, left: `2.5%`, zIndex: 900, width: `95%`, minHeight: `100px` }}>                
                <ProjectForm 
                  type="filter"
                  label="Filter"
                  clearFilter={clearFilter}
                  {...formikProps}              
                />
              </Card>  
            );
          }} 
        /> 
      );
    }
    return null;
  }

       
  
}

/* istanbul ignore next */
const mapStateToProps = state => {
  return {
    projectFilters: state.analyze.projectFilters,
    projectFilterVisible: state.analyze.projectFilterVisible,
    categories: categorySelectors.getAllCategories(state),
    subcategories: subcategorySelectors.getAllSubcategories(state),
    tasks: taskSelectors.getAllTasksWithContent(state)   
  };
};

/* istanbul ignore next */
const mapDispatchToProps = dispatch => {
  return {
    updateFilter: filters => dispatch(analyzeActions.updateFilter(domain.PROJECT,filters)),
    clearFilter: () => dispatch(analyzeActions.clearFilter(domain.PROJECT))
  };
};

export default connect(mapStateToProps,mapDispatchToProps)(ProjectFilter);