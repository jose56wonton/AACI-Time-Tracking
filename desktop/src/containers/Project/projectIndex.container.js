import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { analyzeActions } from 'store/actions';
import { projectSelectors, projectTaskSelectors } from 'store/selectors';
import VirtualizedSortSelect from 'components/tables/VirtualizedSortSelect';
import Progress from 'components/helpers/Progress';
import * as TableDataTypes from 'constants/tableDataTypes';
import { analyzeStatus } from 'constants/analyze';
import domain from 'constants/domains';

export class ProjectIndex extends Component {
  selectLabel = selected =>`${selected.name} selected`;

  select = object => this.props.select(domain.PROJECT,object)

  add = () => this.props.setStatus(domain.PROJECT,analyzeStatus.ADDING)

  render () {
    const { projects,selected } = this.props;

    console.log(`project analyze render`);
    if (!projects) return <Progress variant="circular" fullWidth fullHeight />;

    return (
      <VirtualizedSortSelect 
        data={projects || []}
        columns={rows}
        selected={selected}
        select={this.select}       
        initialSortBy='date'
      />
    );
  }
}

ProjectIndex.propTypes ={ 
  projects: PropTypes.array,
  select: PropTypes.func.isRequired,
  setStatus: PropTypes.func.isRequired,
  selected: PropTypes.object
};

/* istanbul ignore next */
const mapStateToProps = state => {
  const filters = state.analyze.projectFilters;
  return {
    projects: projectSelectors.getAllProjectsNew(state, { filters, sorts: {} }),
    selected: projectTaskSelectors.getSelectedProject(state)
  };
};

/* istanbul ignore next */
const mapDispatchToProps = dispatch => {
  return {
    ...bindActionCreators({ ...analyzeActions }, dispatch)   
  };
};


export default connect(mapStateToProps,mapDispatchToProps)(ProjectIndex);

const rows = [
  {
    id: `name`,  
    dataKey: `name`, 
    width: 150, 
    height: 56,  
    padding: `dense`,
    label: `Name`,
    type: TableDataTypes.STRING
  }, 
  {
    id: `date`,    
    dataKey: `date`, 
    width: 80, 
    height: 56,
    padding: `dense`,
    label: `Date`,
    type: TableDataTypes.DATE
  }, 
  {
    id: `isActive`, 
    dataKey: `isActive`,  
    width: 80, 
    height: 56, 
    align: `right`, 
    padding: `dense`,
    label: `Active`,
    type: TableDataTypes.BOOLEAN
  }   
];
