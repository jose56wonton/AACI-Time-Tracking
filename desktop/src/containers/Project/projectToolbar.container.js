import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { analyzeActions } from 'store/actions';
import { projectSelectors, projectTaskSelectors } from 'store/selectors';
import Progress from 'components/helpers/Progress';
import { analyzeStatus } from 'constants/analyze';
import domain from 'constants/domains';

import AnalyzeToolbar from 'components/tables/toolbars/NewAnalyzeToolbar';

export class ProjectIndex extends Component {
  selectLabel = selected =>`${selected.name} selected`;

  select = object => this.props.select(domain.PROJECT,object)

  add = () => {
    const { selected, select, setStatus } = this.props;
    if(selected && selected.id){
      select(domain.PROJECT,selected.id);
    }
    setStatus(domain.PROJECT,analyzeStatus.ADDING);
  }

  render () {
    const { projects, selected, toggleProjectFilter,projectFilterVisible } = this.props;

    if (!projects) return <Progress variant="circular" fullWidth fullHeight />;

    return (
      <AnalyzeToolbar 
        selectLabel={this.selectLabel}
        label="Projects"
        add={this.add}
        selected={selected}
        toggleFilter={toggleProjectFilter}
        isFilterVisible={projectFilterVisible}
      />
    );
  }
}

ProjectIndex.propTypes ={ 
  projects: PropTypes.array,
  select: PropTypes.func.isRequired,
  setStatus: PropTypes.func.isRequired,
  selected: PropTypes.object,
  toggleProjectFilter: PropTypes.func,
  projectFilterVisible: PropTypes.bool
};

/* istanbul ignore next */
const mapStateToProps = state => {
  return {
    projectFilterVisible: state.analyze.projectFilterVisible,
    projects: projectSelectors.getAllProjects(state),
    selected: projectTaskSelectors.getSelectedProject(state)
  };
};

/* istanbul ignore next */
const mapDispatchToProps = dispatch => {
  return {
    ...bindActionCreators({ ...analyzeActions }, dispatch),
    toggleProjectFilter: () => dispatch(analyzeActions.toggleFilter(domain.PROJECT)) 
  };
};

export default connect(mapStateToProps,mapDispatchToProps)(ProjectIndex);

