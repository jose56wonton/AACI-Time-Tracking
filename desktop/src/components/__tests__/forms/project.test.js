import React from 'react';
import { shallow } from 'enzyme';
import { FieldArray } from 'formik';

import { Project } from 'components/forms/Project/project';
import ProjectHOC from 'components/forms/Project';

const props =  {  
  classes: {},
  label: `label`,
  type: `edit`,
  removeProject: jest.fn(),
  categories: [{ id: 0 },{ id: 1 },{ id: 2 }],
  subcategories: [{ id: 0 },{ id: 1 },{ id: 2 }],
  tasks: [{ id: 0,category: { id: 0,name: `name0` }, subcategory: { id: 0,name: `name0` } },{ id: 1,category: { id: 1,name: `name1` }, subcategory: { id: 1,name: `name1` } },{ id: 2,category: { id: 2,name: `name2` }, subcategory: { id: 2,name: `name2` } }],
  isSubmitting: true,
  resetForm: jest.fn(),
  initialValues: {},
  values: {
    projectTasks: [{ id: 0, categoryId: 0, subcategoryId: 0,taskId:0 },{ id: 1, categoryId: 1, subcategoryId: 1,taskId:1 }]
  },
  errors: {}
};

const renderProps = { 
  remove: jest.fn(),
  push: jest.fn(),
  form: {
    setFieldValue: jest.fn()
  }
};

const setup = overRides => {  
  return shallow(<Project {...props} {...overRides}/>);    
};

const setupHOC = overRides => {
  return shallow(<ProjectHOC {...props} {...overRides}/>);
};

const setupWithRender = overRides => {
  const wrapper = setup();  
  const Render = wrapper.find(FieldArray).first().prop(`render`);  
  return shallow(<Render {...renderProps} {...overRides} />);
};

describe(`Project Component`, () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it(`should render correctly`, () => {
    setup();
  });
  it(`should render correctly withStyles`, () => {
    setupHOC();    
  });
  it(`should call render of projectTask fieldarray`, () => {
    setupWithRender();
  });

  it(`should call onChange Category Field`, () => {
    const wrapper = setupWithRender();
    expect(renderProps.form.setFieldValue).toHaveBeenCalledTimes(0);
    wrapper.find(`#category-field-1`).first().simulate(`change`);
    expect(renderProps.form.setFieldValue).toHaveBeenCalledTimes(2);
    expect(renderProps.form.setFieldValue).toHaveBeenLastCalledWith(`projectTasks.1.taskId`,-1);
  });
  it(`should call onChange Category Field`, () => {
    const wrapper = setupWithRender();
    expect(renderProps.form.setFieldValue).toHaveBeenCalledTimes(0);
    wrapper.find(`#subcategory-field-1`).first().simulate(`change`);
    expect(renderProps.form.setFieldValue).toHaveBeenCalledTimes(1);
    expect(renderProps.form.setFieldValue).toHaveBeenCalledWith(`projectTasks.1.taskId`,-1);
  });
  it(`should remove an projectTask when a remove-projectTask-1`, () => {
    const wrapper = setupWithRender();
    expect(renderProps.remove).toHaveBeenCalledTimes(0);
    wrapper.find(`#remove-projectTask-1`).first().simulate(`click`);
    expect(renderProps.remove).toHaveBeenCalledTimes(1);
    expect(renderProps.remove).toHaveBeenCalledWith(1);
  });
  it(`should remove an projectTask when a remove-projectTask-1`, () => {
    const wrapper = setupWithRender();
    expect(renderProps.push).toHaveBeenCalledTimes(0);
    wrapper.find(`#add-projectTask`).first().simulate(`click`);
    expect(renderProps.push).toHaveBeenCalledTimes(1);
    expect(renderProps.push).toHaveBeenCalledWith({
      categoryId: -1,
      subcategoryId: -1,
      taskId: -1,
      quantity: 1,
      estimateTime: 1
    });
  });
  it(`should call resetForm on Reset button click`, () => {
    const wrapper = setup();
    expect(props.resetForm).toHaveBeenCalledTimes(0);
    wrapper.find(`#project-reset-button`).first().simulate(`click`);
    expect(props.resetForm).toHaveBeenCalledTimes(1);
    expect(props.resetForm).toHaveBeenCalledWith({});
  });
});