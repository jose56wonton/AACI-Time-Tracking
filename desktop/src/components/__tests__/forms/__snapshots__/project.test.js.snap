// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Component should render correctly 1`] = `
<Form>
  <WithStyles(Grid)
    container={true}
    spacing={24}
  >
    <WithStyles(Grid)
      className=""
      item={true}
      xs={12}
    >
      <WithStyles(Typography)
        variant="h6"
      >
        label
      </WithStyles(Typography)>
      <WithStyles(Tooltip)
        title="Delete"
      >
        <WithStyles(IconButton)
          aria-label="Delete"
          onClick={[Function]}
        >
          <pure(DeleteIcon) />
        </WithStyles(IconButton)>
      </WithStyles(Tooltip)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <C
        component={[Function]}
        helper="normal"
        label="Task Name"
        margin="none"
        name="name"
        type="search"
      />
      <C
        component={[Function]}
        helper="normal"
        label="Start Date"
        margin="none"
        name="date"
        type="date"
      />
      <C
        component={[Function]}
        label="Active"
        name="isActive"
      />
    </WithStyles(Grid)>
    <C
      name="projectTasks"
      render={[Function]}
    />
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <WithStyles(Typography)
        color="error"
        variant="button"
      />
      <div>
        <WithStyles(Button)
          color="primary"
          disabled={true}
          type="submit"
          variant="contained"
        >
          Save
        </WithStyles(Button)>
        <WithStyles(Button)
          color="secondary"
          disabled={true}
          id="project-reset-button"
          onClick={[Function]}
          variant="text"
        >
          Reset
        </WithStyles(Button)>
      </div>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</Form>
`;

exports[`Project Component should render correctly withStyles 1`] = `
<Project
  categories={
    Array [
      Object {
        "id": 0,
      },
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
    ]
  }
  classes={
    Object {
      "button": "Project-button-7",
      "error": "Project-error-6",
      "field": "Project-field-4",
      "gridContainer": "Project-gridContainer-1",
      "headerRow": "Project-headerRow-3",
      "row": "Project-row-2",
      "switchBox": "Project-switchBox-5",
    }
  }
  errors={Object {}}
  initialValues={Object {}}
  isSubmitting={true}
  label="label"
  removeProject={[Function]}
  resetForm={[Function]}
  subcategories={
    Array [
      Object {
        "id": 0,
      },
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
    ]
  }
  tasks={
    Array [
      Object {
        "category": Object {
          "id": 0,
        },
        "id": 0,
        "subcategory": Object {
          "id": 0,
        },
      },
      Object {
        "category": Object {
          "id": 1,
        },
        "id": 1,
        "subcategory": Object {
          "id": 1,
        },
      },
      Object {
        "category": Object {
          "id": 2,
        },
        "id": 2,
        "subcategory": Object {
          "id": 2,
        },
      },
    ]
  }
  type="edit"
  values={
    Object {
      "projectTasks": Array [
        Object {
          "categoryId": 0,
          "id": 0,
          "subcategoryId": 0,
          "taskId": 0,
        },
        Object {
          "categoryId": 1,
          "id": 1,
          "subcategoryId": 1,
          "taskId": 1,
        },
      ],
    }
  }
/>
`;
